hg38 support

vinjana opened the pull request:
* 35.1.0 (upcoming)
  * Minor: Generic assembly/reference support
    * Added `--assemblyname` option, defaulting to SOPHIA 35.0.0's chromosome converter implementation "classic_hg37" when omitted.
    * Added "hs37d5+phix" configuration for the generic re-implementation of the "classic_hg37" chromosome converter.
    * Added "hg38" configuration for the generic implementation.
    > WARNING: hg38 support was not excessively tested. In particular, yet hardcoded parameters may have to be adjusted. Furthermore, the runtime will be longer than for "classic_hg37" and also "classic_hg37" runtime has increase slightly (due to class polymorphism).
  * Minor: Build system
    * Use `make` as build system
    * `Release_*` directories with old build-scripts removed
    * Allow static building with `make static=true boost_lib_dir=/path/to/boost/lib`
    * Allow development build with `make develop=true`
    * Build `sophiaMref` (no build documentation before)
    * Build `testRunner` for running unit tests
  * Patch: A `README.md` file that is worth its name and contains first documentation about the usage of the SOPHIA binaries and input and output files.
  * Patch: Added unit tests.
  * Patch: Code readability improvements, documentation, `.editorconfig` file, and `clang-format` configuration
  * Patch: Major refactorings for code clarity (and understanding of the convoluted code) and to improve usage of C++ type system for compiler-based checks of changes.
  * Patch: For `sophiaAnnotate` the default value for clonalitylofreq was advertised in the usage information as 10, but the actual value was 5. Now, the previously used values (5) is also advertised as default.
  * Patch: Performance improvement of `sophiaMref` by avoiding repeated vector-reallocation up to target size 304 GB.

Closes #4, #5, #6, #13 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Enhanced SOPHIA tool with detailed documentation, including usage, build instructions, and support for generic assembly.
  - Introduced new classes and methods for improved functionality and clarity in handling chromosomes, breakpoints, and alignments.
  - Added unit tests for validating functionality across various components.
- **Bug Fixes**
  - Refined logic and fixed issues in handling supplementary alignments and chromosome information.
- **Documentation**
  - Updated README.md with comprehensive tool descriptions, dependencies, and build system instructions.
- **Refactor**
  - Major refactorings across several classes to improve code readability, maintainability, and performance.
  - Standardized naming conventions and updated data types for consistency and clarity.
- **Chores**
  - Updated .gitignore patterns and Makefile configurations for enhanced project management.
- **Style**
  - Improved code formatting for better readability and adherence to best practices.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

vinjana committed:
Made ChrConverter more flexible (second hard-coded reference) and centralized in GlobalAppConfig.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz.de>

vinjana committed:
Build script for sophiaMref.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz.de>

vinjana committed:
Made README more descriptive with texts from Umut's and the SophiaWorkflow readmes.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz.de>

vinjana committed:
More code layout changes. Removed temporary file.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz.de>

vinjana committed:
Removed using namespace std. More code layout changes.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz.de>

vinjana committed:
Documentation for hg37 chromosome name parser.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz.de>

vinjana committed:
Replace obscure -2 into NA variable to get more self-descriptive code.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz.de>

vinjana committed:
Major refactoring to have a map-based hg38 index/converter.

* Added Makefile to replace build scripts.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz.de>

vinjana committed:
Fixed some compile instructions for static compile.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz.de>

vinjana committed:
Compile fix for static libraries with Makefile.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz.de>

vinjana committed:
Fixed many errors, but not all yet.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz.de>

vinjana committed:
Fixed all apparent compile errors. Some semantic errors as warning remain.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz.de>

vinjana committed:
Made compressedMrefIndexToIndex() return an optional to account for failing index mapping.

It is unclear, why the mapping should fail. The optional guarantees that the calling code has to handle the situation. There are many places where the calling code implements a default action, and other placess (e.g. AnnotationProcessor.cpp:227) where an error should occur (used to be segmentation fault, now is exception).

Signed-off-by: Philip R. Kensche <p.kensche@dkfz.de>

vinjana committed:
Makefile simplified. Static compile still doesn't work, though.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz.de>

vinjana committed:
Fully implemented dynamic and static comilation with Makefile. Release_* directories removed.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz.de>

vinjana committed:
Merge pull request #7 from DKFZ-ODCF/static-compile

Static compile with Makefile

NagaComBio commented on .gitignore:
The file needs to be updated with the latest Makefile structure
```
build/*.o
include/strtk.hpp
sophia
sophiaAnnotate
sophiaMref
boost/
```

vinjana replied:
Done.

vinjana committed:
Encapsulate implementation detail into ChrConverter.isIgnoredChromosome(ChrIndex).

* This replace the dubious `chrIndex < 1002` and `chrIndex > 1001` meaning "not p* or INVALID" or "p* or INVALID", with `p*` actually meaning phiX174 in Hg37ChrConverter.
* Wrap the strtk.hpp for C++11 into strkt-wrap.h that turns off all warnings for strtk.hpp, because we use C++17 for the rest of the code.
* Further Makefile improvements.
* More code-comments and small refactorings to improve readability.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz.de>

vinjana commented on include/GlobalAppConfig.h:
```diff
@@ -0,0 +1,68 @@
+        static GlobalAppConfig &init(unique_ptr<ChrConverter const> chrConverter);
+
+        /** Getter. */
+        static const GlobalAppConfig &getInstance();
```
Done.

vinjana commented on src/Hg37ChrConverter.cpp:
```diff
@@ -0,0 +1,375 @@
+            }
+        }
+        return 0;
+    }
```
Has been changed, as far as possible. The Hg37ChrConverter will probably remain like that to keep the old behaviour.

vinjana commented on src/Hg38ChrConverter.cpp:
```diff
@@ -0,0 +1,375 @@
+            case 'p':
+                return 1002;
+            default:
+                return 1003;
```
The case was removed. Now throws.

vinjana commented on src/sophia.cpp:
```diff
@@ -35,83 +43,118 @@ int main(int argc, char** argv) {
+
 std::pair<double, double> getIsizeParameters(const std::string &ISIZEFILE) {
 	std::pair<double, double> isizeMedianStd { };
 	std::ifstream infile { ISIZEFILE };
```
Not gonna happen soon.

vinjana commented on include/ChrConverter.h:
```diff
@@ -19,65 +13,67 @@
-} /* namespace sophia */
-
-#endif /* CHRCONVERTER_H_ */
+    using namespace std;
```
Removed.

vinjana commented on src/Hg38ChrConverter.cpp:
```diff
@@ -0,0 +1,378 @@
+            }
+        }
+        return 0;
+    }
```
Done.

vinjana commented on src/Hg37ChrConverter.cpp:
```diff
@@ -0,0 +1,393 @@
+            }
+        }
+        return 0;
+    }
```
Documented, but won't be changed.

vinjana commented on include/ChrConverter.h:
```diff
@@ -19,65 +13,68 @@
+        virtual ~ChrConverter();
+
+        /** The name of the assembly. */
+        static const std::string assemlyName;
```
fixed

vinjana commented on src/Alignment.cpp:
```diff
@@ -33,680 +34,731 @@
+        } else {
+            mateChrIndex = GlobalAppConfig::getInstance().getChrConverter().parseChrAndReturnIndex(
+                next(samLine.cbegin(), 1 + samChunkPositions[5]), '\t');
+        }
```
Won't be change any time soon.

vinjana committed:
Removed unnecessary imports. Wrap main functions into try-catch to catch and print exception messages.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz.de>

vinjana commented on src/Hg37ChrConverter.cpp:
```diff
@@ -0,0 +1,444 @@
+            if (indexToChrCompressedMref.size() != chrSizesCompressedMref.size())
+                throw std::invalid_argument(
+                    "indexToChrCompressedMref and chrSizesCompressedMref must have the same size");
+        }
```
Done.

vinjana commented on src/Hg37ChrConverter.cpp:
```diff
@@ -0,0 +1,444 @@
+                                        + std::to_string(*startIt) + "'.");
+            }
+        }
+        throw std::runtime_error("Oops! This should not occur!");
```
The code is fine. The final throw is a guard against incorrect changes.

vinjana commented on src/ChrConverter.cpp:
```diff
@@ -19,290 +13,16 @@
-    -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2,
-    -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2,
-    -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2,
-    -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, 83, 84, -2, -2, -2};
+    ChrConverter::~ChrConverter() {}
 
-} /* namespace sophia */
+}
```
That's fine. I just importing `sophia` in the main functions. Nothing more.

vinjana commented on src/Hg37ChrConverter.cpp:
```diff
@@ -0,0 +1,466 @@
+            NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA,
+            NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA,
+            NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA,
+            NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, 83, 84, NA, NA, NA};
```
Will probably come next.

vinjana commented on src/AnnotationProcessor.cpp:
```diff
@@ -217,8 +223,8 @@ AnnotationProcessor::AnnotationProcessor(
         }
-        auto chrIndex = ChrConverter::indexConverter[tmpBp.getChrIndex()];
-        controlResults[chrIndex].push_back(tmpBp);
+        auto chrIndexO = chrConverter.compressedMrefIndexToIndex(tmpBp.getChrIndex());
+        controlResults[chrIndexO.value()].push_back(tmpBp);
```
Not necessary. The context indicates it won't be empty, at least is was implemented like that. We rather throw an exception and fix a bug then mess around too much with the hard business logic in the moment.

vinjana commented on src/sophiaMref.cpp:
```diff
@@ -13,60 +13,97 @@
+	MrefEntry::NUMPIDS = gzListIn.size();
+	MasterRefProcessor mRefProcessor { gzListIn, outputRoot, version, defaultReadLength };
+
+    return 0;
```
Bullshit. The `MasterRefProcessor` does (unfortunately) all the work in the constructor. I won't change that now. There is no `process()` method.

vinjana commented on src/MasterRefProcessor.cpp:
```diff
@@ -35,154 +36,145 @@
+        for (std::vector<int>::size_type i = 0;
+             i < chrConverter.nChromosomesCompressedMref();
+             ++i) {
+            //		mrefDbPtrs.emplace_back(chrSizes[i] + 1, nullptr);
```
done

vinjana commented on src/MasterRefProcessor.cpp:
```diff
@@ -35,154 +36,145 @@
+            chrono::seconds diff =
+                chrono::duration_cast<chrono::seconds>(end - start);
+            ++fileIndex;
+            cerr << gzFile << "\t" << diff.count() << "\t" << newBreakpoints << "\t"
+                 << fileIndex << "\t" << 100 * (fileIndex + 0.0) / NUMPIDS << "%\n";
```
Who knows why `cerr` is used here? The original author might know. In any case, changing the stream might change the output in unintended ways.

vinjana commented on src/Hg37ChrConverter.cpp:
```diff
@@ -0,0 +1,477 @@
+            "985",        "986",        "987",        "988",        "989",
+            "990",        "991",        "992",        "993",        "994",
+            "995",        "996",        "997",        "998",        "hs37d5",
+            "NC_007605",  "MT",         "phiX174",    "INVALID"};
```
It was not extended. It's still the same as before. In the `Hg37ChrConverter` it won't be changed. Reading from file is prepared by the changes in `Hg38ChrConverter.`

vinjana commented on src/Hg37ChrConverter.cpp:
```diff
@@ -0,0 +1,477 @@
+                                            + std::to_string(*start) + "'.");
+            }
+        }
+        throw std::runtime_error("Oops! This should not occur!");
```
That is exactly what was done for hg38.

vinjana commented on include/global.h:
```diff
@@ -0,0 +1,37 @@
+ *     You should have received a copy of the GNU General Public License
+ *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *     LICENSE: GPL
+ */
+
+#ifndef GLOBALAPPCONFIG_H_
+#define GLOBALAPPCONFIG_H_
```
done

vinjana commented on include/ChrConverter.h:
```diff
@@ -19,65 +13,75 @@
-#ifndef CHRCONVERTER_H_
-#define CHRCONVERTER_H_
-#include <array>
+#ifndef _CHRCONVERTER_H_
+#define _CHRCONVERTER_H_
```
done

vinjana commented on src/SuppAlignment.cpp:
```diff
@@ -80,14 +90,17 @@ SuppAlignment::SuppAlignment(string::const_iterator saCbegin, string::const_iter
 	}
-//cerr << "guessSupplementOffset" << endl;
+
+    //cerr << "guessSupplementOffset" << endl;
```
done

vinjana committed:
Fixes after coderabbit review.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz.de>

vinjana committed:
Switched to parse chromosome name function taking start and end iterator and stop-char. Implemented first version of hg38 name parser.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz.de>

vinjana committed:
Fixed input check and improved error message. Changed `make STATIC=true` to `make static=true`. Added `make develop=true` for debugging and profiling.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz.de>

vinjana committed:
Fixed cornercase with supplementary alignment tag (SA:Z:) containing a perfect match (e.g. 101M) in the CIGAR string.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz.de>

vinjana committed:
Minor.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz.de>

vinjana committed:
Added comment.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz.de>

vinjana committed:
Improve chrName parser for breakpoint files, to be able to deal with chromosomes containing colons ':'

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Added googletest/gtest-based unit tests for the chromosome name parser and fixed it.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Fixed Makefile to build static binaries, except the testRunner (which is sufficient to be linked dynamically).

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Fixed wrong initialiazation order of GlobalAppConfig.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Bugs fixed, but parser of breakpoints still broken.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Added boost::stacktrace for better error reporting.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Documented output BED file and code.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Added failing Breakpoint parser tests.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Hack to prevent failure in tests due to reinitialized singleton.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Added documentation for sophiaMref output file. Added `binaries` target and improved compilation docs.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Added tests. Removed TODO.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Added contig classes, also to hg38 (not working yet). Range approach.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Migrated Hg38ChrConverter to one that reads configuration file. Yet hardcoded. Added lots of refactorings for clarity tests.

* New dependency on rapidcsv including download with `make`.
* Explicit chromosome categories make code clearer.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Transformed the Hg38ChrConverter into a simple GenericChrConverter that is easy to configure for development purposes (not production).

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Changed from hg37 to classic_hg37 as default Hg37ChrConverter.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Removed obsolete IndexRange. Fixed is$category methods for classic_hg37. Fixed printing stacktrace. "backtrace" as new dependency!

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Much improved error messages for parsing.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Some minor exception tweaks.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Fixed a bug I intruduced. Noted that SuppAlignmentAnno::SuppAlignmentAnno(const string&) and SuppAlignment::parseSaSupport(const string&) are almost identical and parse the identical syntax.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Comments and removed strange 'm' printed by `sophiaAnnotate`.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Added CONTRIBUTORS.md.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
README update. Added hs37d5+phix.tsv to resources.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Fix chromosome label.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Fixed some chromosome index mappings that could have been entirely avoided, if proper typing would be used.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Major refactorings. Have unsigned and signed int for ChrIndex and CompressedMrefIndex + -Wsign-conversion -Werror to ensure the to types are different.

The strict integer checks, however, required that I worked through basically all code to check the (previously automatic) conversions.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Commandline fixes created by refactoring.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana commented on src/sophia.cpp:
```diff
@@ -35,89 +44,120 @@ int main(int argc, char** argv) {
+	while (error_terminating_getline(infile, line)) {
 		boost::algorithm::trim_right(line);
 		switch (i) {
 		case 0:
```
True, but won't be changed.

vinjana commented on include/GlobalAppConfig.h:
```diff
@@ -0,0 +1,68 @@
+        static GlobalAppConfig &init(std::unique_ptr<ChrConverter const> chrConverter);
+
+        /** Getter. */
+        static const GlobalAppConfig &getInstance();
```
No. Global state should not be modified. It's already messy enough.

vinjana commented on include/GlobalAppConfig.h:
```diff
@@ -0,0 +1,68 @@
+
+        GlobalAppConfig(std::unique_ptr<ChrConverter const> chrConverter);
+
+        ~GlobalAppConfig();
```
done

vinjana commented on src/sophia.cpp:
```diff
@@ -35,83 +43,119 @@ int main(int argc, char** argv) {
+	Breakpoint::BPSUPPORTTHRESHOLD = bpSupport;
+	Breakpoint::DEFAULTREADLENGTH = defaultReadLength;
+	Breakpoint::DISCORDANTLOWQUALLEFTRANGE = static_cast<int>(round(defaultReadLength * 1.11));
+	Breakpoint::DISCORDANTLOWQUALRIGHTRANGE = static_cast<int>(round(defaultReadLength * 0.51));
```
This won't be changed in this PR.

vinjana commented on src/SuppAlignment.cpp:
```diff
@@ -135,194 +250,172 @@ SuppAlignment::SuppAlignment(string::const_iterator saCbegin, string::const_iter
+	if (properPairErrorProne) {
+		outStr.append("#");
+	}
+	return outStr;
```
Thu method was actually not changed much.

vinjana commented on src/SuppAlignment.cpp:
```diff
@@ -135,194 +250,172 @@ SuppAlignment::SuppAlignment(string::const_iterator saCbegin, string::const_iter
+			            (pos - fuzziness) <= (rhs.getExtendedPos() + fuzziness);
 		} else {
 			return abs(pos - rhs.getPos()) <= fuzziness;
 		}
```
No, only the code layout was changed.

vinjana committed:
Switched types for ChrIndex and CompressedMrefIndex and fixed errors. Added coding convention file.

The GenericChrConverter still does not produce the correct results for hg37.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Comments and layout changes.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana commented on src/SamSegmentMapper.cpp:
```diff
@@ -26,456 +26,467 @@
+    //<< endl; 	*metaOutputHandle << "#Using (-F 0x600 -f 0x001)" << endl;
+    //	*metaOutputHandle << "#done\t" << printedBps << " lines printed in " <<
+    //elapsedTime.quot << " minutes, " << elapsedTime.rem << " seconds" << endl;
+    // }
```
old comment. ignored

vinjana commented on src/MrefEntry.cpp:
```diff
@@ -20,215 +20,228 @@
+                return true;
+            }
+        }
+        return false;
```
No feature changed.

vinjana commented:
@coderabbit resolve

vinjana commented:
@coderabbit pause

vinjana commented on include/ChrConverter.h:
```diff
@@ -19,65 +13,134 @@
+        virtual ~ChrConverter();
+
+        /** The name of the assembly. */
+        static const std::string assemblyName;
```
Nope. For the time being, the whole assembly configuration will be static.

vinjana committed:
Reviewed my own code, to find any bugs.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana commented on src/AnnotationProcessor.cpp:
```diff
@@ -491,31 +540,32 @@ void
                                           vector<vector<MrefEntryAnno>> &mref,
-                                          bool doubleSupportSa) {
+                                          bool doubleSupportSa[[gnu::unused]] // TODO: remove
+                                          ) {
```
Not during this mere refactoring. Only if we start thinking about the actual algorithm.

vinjana commented on src/AnnotationProcessor.cpp:
```diff
@@ -542,21 +592,24 @@ AnnotationProcessor::checkSvQuality() {
 AnnotationProcessor::searchMrefHitsNew(const BreakpointReduced &bpIn,
                                        int distanceThreshold,
-                                       int conservativeDistanceThreshold,
+                                       int conservativeDistanceThreshold[[gnu::unused]], // TODO: remove
```
dito

vinjana committed:
Added assertions (can be turned off with -DNDEBUG for production) to check indices. Some smaller refactorings.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Refined tests for valid chromosome indices (classic_hg37), at construction time, and at runtime, base on the static construction time methods.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Fixed check.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Small refactorings

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Refactored ChrCategory, wrote some tests, and fixed some issues.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Changed all index and position types back to int-length. MrefEntry and via that, also SuppAlignmentAnno are copied 3.7 billion times into memory during sophiaMref.

Changed a lots of C-style casts into C++ style casts (static_cast).

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Changed MrefEntry::validity into signed char, to reduce space requirements.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Little refactoring for code readability.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Fixed the memory issue.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Removed all assertions, to see whether that fixes the memory issue.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Small changes.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Added test for internal static constructor function of Hg37ChrConverter.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Removed assertValid calls completely. Instead now use IndexRange class.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Minor.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Preallocate full memory for MasterMrefProcessor (a lot), rather than using emplace_back.

This will eventually reserve less memory, and avoid repeated reallocation of gigabytes of memory.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Some small refactorings, edits, and comments.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Added tests for SuppAlignmentAnno.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Some debugging output to catch a specific difference

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Fixed some incorrectly translated conditions concerning decoys.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Split gonosomes into X and Y classes. One condition (CompressedMrefIndex < 23) actually includes autosomes and X chromosome, but excludes the Y chromosome.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Made assemblyName a non-static, const value in ChrConverter.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Calculate memory allocated by MasterRefProcessor and report in GB.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Minor.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana commented on src/SuppAlignment.cpp:
```diff
@@ -338,7 +431,8 @@ bool SuppAlignment::saDistHomologyRescueCloseness(const SuppAlignment& rhs, int
+			            (pos - fuzziness) <= (rhs.getExtendedPos() + fuzziness);
 		} else {
 			return abs(pos - rhs.getPos()) <= fuzziness;
 		}
```
yes please

vinjana commented on src/Hg37ChrConverter.cpp:
```diff
@@ -0,0 +1,570 @@
+            }
+        } else {
+            switch (*start) {
+                case 'h':
+                    chrIndex = hg37::decoyIndex;
+                    break;
+                case 'X':
+                    chrIndex = hg37::xIndex;
+                    break;
+                case 'G':   // Match GL...... chromosomes.
+                    for (auto cit = next(start, 2); *cit != '.'; ++cit) {
+                        chrIndex = 10 * chrIndex + ChrIndex(*cit - '0');
+                    }
+                    break;
+                case 'Y':
+                    chrIndex = hg37::yIndex;
+                    break;
+                case 'M':   // Match "MT"
+                    ++start;
+                    if (start != end && *start == 'T') {
+                        chrIndex = hg37::mtIndex;
+                    } else {
+                        throw_with_trace(
+                            DomainError("Chromosome identifier with invalid prefix 'M" +
+                                              std::to_string(*start) + "'."));
+                    }
+                    break;
+                case 'N':
+                    chrIndex = hg37::decoyIndex;
+                    break;
+                case 'p':
+                    chrIndex = hg37::phixIndex;
+                    break;
+                default:
+                    throw_with_trace(DomainError("Chromosome identifier with invalid prefix '"
+                                                       + std::to_string(*start) + "'."));
+            }
+        }
+        return chrIndex;
+    }
```
yes, please

vinjana committed:
Answers to CodeRabbit.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Fixed incorrect index returned from parse function.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Adapted SvEvent conditions again to make them make similar to dealing with CompressedMref chromosomes (that I thought would not occur at that place).

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Removed all using namespace std.

Signed-off-by: Philip R. Kensche <p.kensche@dkfz-heidelberg.de>

vinjana committed:
Merge pull request #22 from DKFZ-ODCF/review-my-code

Restore reference behaviour for sophiaMref

vinjana commented:
Commit e65cfbf produces the same `sophiaMref` results using "classic_hg37" `Hg37ChrConverter` for our test data (@NagaComBio). This means, the previous bugs introduced in the `ChrConverter` client code

Next tests:
1. Check `sophia` tool with "classic_hg37" vs. original `sophia` tool (master)
2. Check `sophiaAnnotate` tool with "classic_hg37" vs. original `sophiaAnnotate` tool (master)

After that we have established, that all client code of the `ChrConverter` API remained intakt.

Then we can check `sophiaMref`, `sophia`, and `sophiaAnnotate` of the "hg37" (=`GenericChrConverter`) against the reference results. Then we know that the `GenericChrConverter` reimplements (in a more generic way) exactly the behavior of the original `Hg37ChrConverter`.

After that we can use "hg38", which is also implemented with the `GenericChrConverter`.
